### PR TITLE
DIFM: Manage Purchases: Show extra page price breakdown using price tiers

### PR DESCRIPTION
**Pull request by ddc22**
#### Proposed Changes

Adds a proper description and cost breakdown for DIFM purchase receipt view.

#### Testing Instructions
- Apply  D85709-code
- Apply D86680-code
- Sandbox Public API
- Use the below calypso live link go to `/start/do-it-for-me`
- Go through the flow and select more than 5 pages in the page picker step
- Complete purchase
- Go to purchases `http://calypso.localhost:3000/purchases/subscriptions/<site-slug>`
- Select the DIFM purchase and make sure you see below
image

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Fixes 937-gh-Automattic/martech

### CODE REVIEW

File: client/me/purchases/manage-purchase/index.jsx [hunk 1]
```diff
@@ -642,10 +643,42 @@ class ManagePurchase extends Component {
...
-
 			return description;
 		}
 
+		if ( isDIFMProduct( purchase ) ) {
+			const productDetails = productsList[ WPCOM_DIFM_LITE ];
+			const [ tier0 ] = productDetails.price_tier_list;
+			const { maximum_units: noOfBaselinePagesIncluded } = tier0;
+
+			const description = translate(
```
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/index.jsx]: Minor, but if `description` isn't used when `purchase.purchaseRenewalQuantity` is falsy, we could move this `translate` call inline into return statement in the next block. 🤷

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 2]
```diff
@@ -223,11 +225,51 @@ function PurchaseMetaOwner( { owner } ) {
...
-	const plan = getPlan( productSlug ) || getProductFromSlug( productSlug );
+	const { productDisplayPrice } = purchase;
+	const productsList = useSelector( getProductsList );
 	let period = translate( 'year' );
 
 	if ( isOneTimePurchase( purchase ) || isDomainTransfer( purchase ) ) {
+		if ( isDIFMProduct( purchase ) ) {
+			const productDetails = productsList[ WPCOM_DIFM_LITE ];
+			const [ tier0, tier1 ] = productDetails.price_tier_list;
+			const perExtraPagePrice = ( tier1.minimum_price - tier0.minimum_price ) / 100;
```
Comment (sirbrillig) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: The `100` here is something of a [magic number](https://www.codebyamir.com/blog/software-anti-patterns-magic-numbers). What does it represent? Could we put it in a named variable/constant/function to make that more clear?

The main reason I was wondering is that sometimes I've seen this sort of math used for calculations that don't apply to currencies which do not have a decimal (eg `JPY`) and I just want to make sure it's clear what is happening here.

File: packages/wpcom-checkout/src/can-item-be-removed-from-cart.ts [hunk 3]
```diff
@@ -31,13 +31,13 @@ export function canItemBeRemovedFromCart(
...
-	const itemTypesThatCannotBeDeleted = [ 'domain_redemption' ];
+	const itemTypesThatCannotBeDeleted = [ 'domain_redemption', 'wp_difm_extra_page' ];
 	if ( itemTypesThatCannotBeDeleted.includes( item.product_slug ) ) {
 		return false;
 	}
 
-	// The Premium plan cannot be removed from the cart when in combination with the DIFM lite product
-	if ( isPremiumPlanWithDIFMInTheCart( item, responseCart ) ) {
+	// The Pro plan cannot be removed from the cart when in combination with the DIFM lite product
+	if ( isProPlanWithDIFMInTheCart( item, responseCart ) ) {
```
Comment (sirbrillig) [inline packages/wpcom-checkout/src/can-item-be-removed-from-cart.ts]: The PR title or description doesn't mention anything about changes to what products can be removed from the cart. Neither does the commit. Can you say more about this part of the change?

File: client/me/purchases/manage-purchase/purchase-meta.jsx (hunk 2 above)
Comment (ddc22) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: Thanks @sirbrillig I was actually trying to get the answer I was looking for based on the price that I was familiar with. Thanks for raising this!
Is there a better way to get the actual number to be displayed? Or should I try to resolve this in the backend?
CC - @aneeshd16 @DavidRothstein

File: packages/wpcom-checkout/src/can-item-be-removed-from-cart.ts (hunk 3 above)
Comment (ddc22) [reply packages/wpcom-checkout/src/can-item-be-removed-from-cart.ts]: This was just changing Premium -> Pro which needs to revert back now that we are back on premium plans 😅 . Thanks for raising this!

File: client/me/purchases/manage-purchase/purchase-meta.jsx (hunk 2 above)
Comment (sirbrillig) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: > Is there a better way to get the actual number to be displayed? 

🤔 that's a good question. I'm not saying your technique is wrong, I just don't understand it. 😅  

Let's see, if I understand correctly, you want to get the price per page that the user has already paid for extra DIFM pages attached to this purchase. We know that the total price they paid was [purchase.amount](https://github.com/Automattic/wp-calypso/blob/c8b35df0d209e20a9b6365468cbc894fad320624/client/lib/purchases/types.ts#L3) but the purchases endpoint doesn't appear to include any information on the breakdown of that price. (Maybe that's something that could be added on the backend?) To determine it, we're looking at the price from the products endpoint for the `wp_difm_lite` product and comparing the price between tiers before dividing by 100. I still don't really understand how that gives the price per page, but it makes me wonder: could the price for the `wp_difm_lite_extra_page` be used for this? (Please forgive me if that's totally wrong. I don't really grok the pricing of DIFM.)

Screen Shot 2022-08-23 at 6 10 55 PM

Comment (ddc22) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: Ah yes we are actually moving to a tiered pricing system. This product will be removed.
Context:  pdh1Xd-Oi-p2#comment-1229, D85709-code
So I have to do "math" for this to figure out the per price page. As you said maybe there is a way to get the  the per page price from the backend. Maybe @DavidRothstein will have some advice for us 🤔

Comment (aneeshd16) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: If I understand correctly, we are trying to calculate the cost of the standalone service and the cost of extra pages from the `product` object. However, the product object will change if we update our pricing in the future and will always have the product's current price. So, calculating these prices from the values in the product object will show incorrect prices for earlier purchases. 

We should be calculating this only from the `purchase` object, which (I believe) is immutable. At this time, the `purchase` object returned from the API does not contain the fields necessary to calculate this, so we would need to add these fields on the backend. The `purchase` object should be able to access the billing plan on the backend (also immutable), so this should be doable.

Comment (ddc22) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: Ah right I will have a look into this, thanks for the context @aneeshd16 .
In the mean time we can also consider the approach that you have taken which is somewhat similar but using the `formatCurrency` function with the `isSmallestUnit` flag!

https://github.com/Automattic/wp-calypso/pull/66872/files#diff-d6d276e55159d4aca386ac0391e990943d93b5dc197e6dc3fad4b0840db3f277R660-R665

Comment (aneeshd16) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: Sure, but we will need to make these calculations with values from the purchases API. Using values from the `product` object will become a blocker for future pricing updates.

Comment (ddc22) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: What I meant was just using this function :D @aneeshd16 

```javascript
formatCurrency( price, currency, {
		isSmallestUnit: true,
} );
```

Comment (DavidRothstein) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: >We should be calculating this only from the `purchase` object, which (I believe) is immutable. At this time, the `purchase` object returned from the API does not contain the fields necessary to calculate this, so we would need to add these fields on the backend.

That sounds right to me too.

Basically the `price_tier_list` property that's currently on the `product` object could have an equivalent on the `purchase` object -- which in the server-side code would probably be built inside the subscription's `set_properties_for_purchases_rest_api()` method using `$this->product->plan` (or something similar) to get the billing plan (and then extract the tier data from the billing plan the same way the `product` endpoints do).  That would ensure all your data is coming from the billing plan associated with this particular subscription.

Comment (ddc22) [reply client/me/purchases/manage-purchase/purchase-meta.jsx]: Thanks for the @aneeshd16 @DavidRothstein for the suggestion. I have implemented the backend change as suggested.
D86680-code

File: client/me/purchases/manage-purchase/index.jsx [hunk 4]
```diff
@@ -646,10 +647,42 @@ class ManagePurchase extends Component {
 					</>
 				);
 			}
-
 			return description;
 		}
 
+		if ( isDIFMProduct( purchase ) ) {
+			const productDetails = productsList[ WPCOM_DIFM_LITE ];
+			const [ tier0 ] = productDetails.price_tier_list;
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/index.jsx]: Now that we have added those fields, we can use `price_tier_list` from the `purchase` object.

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 5]
```diff
@@ -20,6 +21,7 @@ import { times } from 'lodash';
 import PropTypes from 'prop-types';
 import { useState, useEffect } from 'react';
 import { useSelector } from 'react-redux';
+import formatCurrency from 'calypso/../packages/format-currency/src';
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: The correct import for this is `@automattic/format-currency`. 🙂

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 6]
```diff
@@ -221,13 +224,72 @@ function PurchaseMetaOwner( { owner } ) {
...
 }
 
+const formatPurchasePrice = ( price, currency ) =>
+	formatCurrency( price, currency, {
+		isSmallestUnit: true,
+		stripZeros: true,
+	} );
+
+function getDIFMPriceDetails( purchase ) {
+	const [ tier0, tier1 ] = purchase.priceTierList;
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: Nice! 👍

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 7]
```diff
@@ -221,13 +224,72 @@ function PurchaseMetaOwner( { owner } ) {
...
+	return {
+		extraPageCount,
+		costOfExtraPages: formattedCostOfExtraPages,
+		oneTimeFee: formattedOneTimeFee,
+	};
+}
+
 function PurchaseMetaPrice( { purchase } ) {
 	const translate = useTranslate();
+	const productsList = useSelector( getProductsList );
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: I believe this is no longer required.

File: client/lib/purchases/assembler.ts [hunk 8]
```diff
@@ -63,6 +63,18 @@ function createPurchaseObject( purchase: RawPurchase | RawPurchaseCreditCard ):
...
+				minimumUnits: ptRaw.minimum_units,
+				maximumUnits: ptRaw.maximum_units,
+				minimumPrice: ptRaw.minimum_price,
+				maximumPrice: ptRaw.maximum_price,
+				minimumPriceDisplay: ptRaw.minimum_price_display,
+				maximumPriceDisplay: ptRaw.maximum_price_display,
+				minimumPriceMonthlyDisplay: ptRaw.minimum_price_monthly_display,
+				maximumPriceMonthlyDisplay: ptRaw.maximum_price_monthly_display,
+			} )
+		),
```
Comment (aneeshd16) [inline client/lib/purchases/assembler.ts]: Suggestion to make this more readable:
```
priceTierList: purchase.price_tier_list?.map( ( tier ) => mapRecordKeysRecursively( tier, snakeToCamelCase ) );
```

File: client/lib/purchases/types.ts [hunk 9]
```diff
@@ -64,6 +65,28 @@ export interface Purchase {
...
+
+export interface PriceTierRaw {
+	minimum_units: number;
+	maximum_units: number;
+	minimum_price: number;
+	maximum_price: number;
+	minimum_price_display: string;
+	maximum_price_display: string;
+	minimum_price_monthly_display: string;
+	maximum_price_monthly_display: string;
```
Comment (aneeshd16) [inline client/lib/purchases/types.ts]: We can remove these since these fields have been removed from the API.

File: client/lib/purchases/types.ts [hunk 10]
```diff
@@ -64,6 +65,28 @@ export interface Purchase {
...
+	maximumPriceMonthlyDisplay: string;
+}
+
+export interface PriceTierRaw {
+	minimum_units: number;
+	maximum_units: number;
+	minimum_price: number;
+	maximum_price: number;
+	minimum_price_display: string;
+	maximum_price_display: string;
```
Comment (aneeshd16) [inline client/lib/purchases/types.ts]: Since `maximum_price` and `maximum_price_display` can be `null`, we should change the type accordingly.
image

File: client/me/purchases/manage-purchase/index.jsx [hunk 11]
```diff
@@ -646,10 +647,42 @@ class ManagePurchase extends Component {
...
+			const { maximum_units: noOfBaselinePagesIncluded } = tier0;
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/index.jsx]: Nit: Maybe rename to `numberOfIncludedPages`? Asking because the `baseline` term might be confusing here.

File: client/me/purchases/manage-purchase/index.jsx [hunk 12]
```diff
@@ -646,10 +647,42 @@ class ManagePurchase extends Component {
...
+
+			if ( purchase.purchaseRenewalQuantity ) {
+				return (
+					<>
+						{ description }{ ' ' }
+						{ translate(
+							'This purchase includes %(numberOfPages)d extra page.',
+							'This purchase includes %(numberOfPages)d extra pages.',
+							{
+								count: purchase.purchaseRenewalQuantity,
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/index.jsx]: The `count` should also be `purchase.purchaseRenewalQuantity - noOfBaselinePagesIncluded`.

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 13]
```diff
@@ -221,13 +224,72 @@ function PurchaseMetaOwner( { owner } ) {
...
+								oneTimeFee,
+							},
+							components: {
+								period: <span className="manage-purchase__time-period" />,
+							},
+						} ) }
+					</div>
+					<div>
+						{ translate(
+							'%(extraPageCount)d extra pages : %(costOfExtraPages)s {{period}}(one-time){{/period}}',
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: We should also add singular/plural versions here, similar to the ones added in `client/me/purchases/manage-purchase/purchase-meta.jsx`.

**Review by aneeshd16 — CHANGES_REQUESTED** [review]
Works well with new purchases. Great job on the backend changes! 👍 

I have left some code comments and some feedback below:

* This change should also handle past purchases. The price for past purchases is shown incorrectly. Maybe we can show the new text only if tier data is available on the purchase object?
image

* The purchase description does seem a bit confusing on this page (Manage Purchases). The user has already "hired" a professional. Maybe this could be something like "Website built by a WordPress.com professional. This build contains 9 pages". What do you think, @mikeicode?
image

File: client/me/purchases/manage-purchase/index.jsx (hunk 4 above)
Comment (ddc22) [reply client/me/purchases/manage-purchase/index.jsx]: Nice catch! :)

File: client/lib/purchases/assembler.ts (hunk 8 above)
Comment (ddc22) [reply client/lib/purchases/assembler.ts]: TIL 👍

Comment (ddc22) [reply client/lib/purchases/assembler.ts]: On second thought I wonder if this is a better approach since all of the type safety is lost. And the type is converted to a more generic Record<string,unknown> type. Going to stick to the existing approach. Let me know if you can think of an approach that can also preserve the types.

File: client/lib/purchases/types.ts [hunk 14]
```diff
@@ -64,6 +67,20 @@ export interface Purchase {
...
+export interface PurchasePriceTier {
+	minimumUnits: number;
+	maximumUnits?: undefined | null | number;
+	minimumPrice: number;
+	maximumPrice: number;
+	minimumPriceDisplay: string;
+	maximumPriceDisplay?: string | null | undefined;
+}
+
+export interface RawPurchasedPriceTierEntry extends PriceTierEntry {
```
Comment (aneeshd16) [inline client/lib/purchases/types.ts]: Nit: This could be `RawPurchasePriceTierEntry` instead of `RawPurchasedPriceTierEntry` to align with the convention in this file.

File: client/lib/signup/step-actions/index.js [hunk 15]
```diff
@@ -321,7 +321,11 @@ export function setThemeOnSite( callback, { siteSlug, themeSlugWithRepo } ) {
 function addDIFMLiteToCart( callback, dependencies, step, reduxStore ) {
 	const { selectedDesign, selectedSiteCategory, isLetUsChooseSelected, siteSlug } = dependencies;
 	const extra = buildDIFMCartExtrasObject( dependencies );
-	const cartItem = { product_slug: WPCOM_DIFM_LITE, extra };
+	const cartItem = {
+		product_slug: WPCOM_DIFM_LITE,
+		extra,
+		quantity: dependencies.selectedPageTitles.length,
```
Comment (aneeshd16) [inline client/lib/signup/step-actions/index.js]: Since we haven't shipped the backend change yet, I think we should move this change to a different PR.

File: client/me/purchases/manage-purchase/index.jsx [hunk 16]
```diff
@@ -646,10 +647,42 @@ class ManagePurchase extends Component {
 					</>
 				);
 			}
-
 			return description;
 		}
 
+		if ( isDIFMProduct( purchase ) ) {
+			const productDetails = productsList[ WPCOM_DIFM_LITE ];
+			const [ tier0 ] = productDetails.priceTierList;
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/index.jsx]: We can use the `priceTierList` from the `purchase` object since it's now available.

Comment (aneeshd16) [reply client/me/purchases/manage-purchase/index.jsx]: This also gives this error since the product details might not have been loaded from the server:
image

Comment (ddc22) [reply client/me/purchases/manage-purchase/index.jsx]: Sorry I thought I took care of this 😬

File: client/lib/signup/step-actions/index.js (hunk 15 above)
Comment (ddc22) [reply client/lib/signup/step-actions/index.js]: Hiding behind feature flag 👍

File: client/lib/purchases/types.ts [hunk 17]
```diff
@@ -64,6 +67,20 @@ export interface Purchase {
...
+export interface PurchasePriceTier {
+	minimumUnits: number;
+	maximumUnits?: undefined | null | number;
+	minimumPrice: number;
+	maximumPrice: number;
+	minimumPriceDisplay: string;
+	maximumPriceDisplay?: string | null | undefined;
+}
+
+export interface RawPurchasdPriceTierEntry extends PriceTierEntry {
```
Comment (aneeshd16) [inline client/lib/purchases/types.ts]: Typo :) Should be `RawPurchasePriceTierEntry`.

File: client/me/purchases/manage-purchase/index.jsx [hunk 18]
```diff
@@ -988,6 +1017,7 @@ function PurchasesQueryComponent( { isSiteLevel, selectedSiteId } ) {
 export default connect(
 	( state, props ) => {
 		const purchase = getByPurchaseId( state, props.purchaseId );
+		const difmTieredPurchaseDetails = getDIFMTieredPurchaseDetails( state, props.purchaseId );
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/index.jsx]: I like the refactor to extract this into a function! Can you please explain why this function needs to fetch the `purchase` object from the state again? The `purchase` object is already available near both usages of the function. The new function is not using `state` for anything other than to fetch the `purchase` object, so it seems redundant to pass in the state when the purchase object is already available. Can we directly pass this object to the new function? This function signature could look like `getDIFMTieredPurchaseDetails( purchase ) => <return value>`.

This new function could live in `client/lib/purchases/index.ts`, where almost all functions have a similar signature. ( `fn( purchase ) => return value`). Admittedly, this file is getting quite large, but since the functions seem to be pure, it should be easy to refactor in the future.

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 19]
```diff
@@ -224,11 +226,54 @@ function PurchaseMetaOwner( { owner } ) {
...
+				} = difmTieredPurchaseDetails;
+				return (
+					<div>
+						<div>
+							{ translate( 'Service : %(oneTimeFee)s {{period}}(one-time){{/period}}', {
+								args: {
+									oneTimeFee,
+								},
+								components: {
+									period: <span className="manage-purchase__time-period" />,
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: This class has no styles in the code base, so maybe we can remove the `span` element here?

**Review by aneeshd16 — COMMENTED** [review]
Tests well! I left some code feedback below.

File: client/me/purchases/manage-purchase/index.jsx (hunk 18 above)
Comment (ddc22) [reply client/me/purchases/manage-purchase/index.jsx]: Actually I was modelling it as a REDUX selector so it can be used anywhere that state slice needs to be calculated. I actually don't understand why it needs to go in a separate lib directory when this is clearly a calculation of existing state 🤔.

Comment (aneeshd16) [reply client/me/purchases/manage-purchase/index.jsx]: The calculation is on the `purchase` object, which could come from different places - it could be from the global redux state (like here), from some other store accessed by `@wordpress/data` ([example usage](https://github.com/Automattic/wp-calypso/blob/93d3b00576d8c8173cfc488d2bf097ff87165fea/client/landing/stepper/declarative-flow/internals/steps-repository/edit-email/index.tsx#L23)), or directly from the results of an API call (since a lot of implementations are moving to `react-query`). The `state` is not used for anything other than to fetch the purchase object. Converting this to a lib function also makes writing automated tests easier.

Also, before calling this function, we are relying on the fact that the code has checked if this is a DIFM purchase. Otherwise, the output of this function does not make sense for non-DIFM purchases.

Comment (ddc22) [reply client/me/purchases/manage-purchase/index.jsx]: > The calculation is on the purchase object, which could come from different places - it could be from the global redux state (like here), from some other store accessed by @wordpress/data ([example usage](https://github.com/Automattic/wp-calypso/blob/93d3b00576d8c8173cfc488d2bf097ff87165fea/client/landing/stepper/declarative-flow/internals/steps-repository/edit-email/index.tsx#L23)), or directly from the results of an API call. The state is not used for anything other than to fetch the purchase object.

Actually the intention of this piece of code was to extract the purchase details from the REDUX state slice which holds the information cached from /purchases endpoint. Nothing else. What your explaining sounds familiar to the [util method anti pattern](https://www.yanglinzhao.com/posts/utils-antipattern/) similar to util class but for functions. Or am I wrong? This function does one thing. Transforms and calculates DIFM related price tier details if available in the identified purchase of the purchases slice of the supplied redux tree and will return null if not.

> Also, before calling this function, we are relying on the fact that the code has checked if this is a DIFM purchase. Otherwise, the output of this function does not make sense for non-DIFM purchases.

Yes the code anticipates that it will be call on any variation of the universal state slice. And it will handle all those variations as appropriate :). We can possibly check the purchase Id to see if it's actually DIFM an reject the result but I don't think that is necessary. If someone wants to reuse this logic for something identical they can refactor the name when they find this function. But for others this encapsulates the price tier's extraction logic for a DIFM purchase so long as you provide the proper purchase ID of a difm purchase you will get the proper details back.

Comment (ddc22) [reply client/me/purchases/manage-purchase/index.jsx]: I added some changes to improve the clarity of the selector 👍 
https://github.com/Automattic/wp-calypso/pull/66559/commits/7a245a208e877779cf262ea7ebf6da4ec8acb9ef

Comment (aneeshd16) [reply client/me/purchases/manage-purchase/index.jsx]: > Transforms and calculates DIFM related price tier details if available in the identified purchase of the purchases slice of the supplied redux tree and will return null if not.

I understand that this function is currently designed to pick the purchase object from the state slice. However, IMO, this function is better suited to work on the `purchase` object itself. My reasoning:
* The function currently does not use any other part of the state to calculate the result. It only picks the finds the `purchase` object again, which we already have access to when we call this function.
* This function is executed for all types of purchases since it is being called in the `connect` function here, which seems redundant to me.
* The function is tightly coupled to the redux state, and assumes that the `purchase` object will always be in the redux state. However, it may not be the case as:
  * The [recommended approach for data is using react-query](https://github.com/Automattic/wp-calypso/blob/trunk/client/data/README.md). For future changes that involve fetching purchase data from the backend, this function will have to be refactored to work on the `purchase` object.
  * Not all data is stored in the global redux state. In the stepper framework for example, the data is stored in `@wordpress/data` stores. ([link](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/README.md))

According to the article you linked to, I don't this is an anti-pattern as the `lib/purchases` directory has a clearly defined boundary (`purchase` object functions) and it is not a part of general app-wide util module.

However, let's keep this PR moving forward 🙂 If we do decide to keep the function as it is, I have added one more bit of feedback for the change in `client/me/purchases/manage-purchase/purchase-meta.jsx`.

File: client/me/purchases/manage-purchase/purchase-meta.jsx [hunk 20]
```diff
@@ -224,11 +226,48 @@ function PurchaseMetaOwner( { owner } ) {
 
 function PurchaseMetaPrice( { purchase } ) {
 	const translate = useTranslate();
+	const overallState = useSelector( ( state ) => state );
```
Comment (aneeshd16) [inline client/me/purchases/manage-purchase/purchase-meta.jsx]: We might as well use `useSelector( state => getDIFMTieredPurchaseDetails( state, purchase.id )` here. Otherwise, this component will be re-rendered if any part of the global state changes.

File: client/me/purchases/manage-purchase/index.jsx (hunk 18 above)
Comment (ddc22) [reply client/me/purchases/manage-purchase/index.jsx]: Oh I wasn't aware of the move to react-query TIL! . I had no idea we were evolving the data layer to something different! 

In which case trying to maintain a set of clean set of selectors is a non starter anyway (which was my main motivation)!
I will move this to the lib folder as you recommended! :)

**Review by aneeshd16 — APPROVED** [review]
LGTM! Great work!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7508317

Thank you @jdc91 for including a screenshot in the description! This is really helpful for our translators.

**Comment by emilyaudela** [comment]
Hi all! Any chance w can add a plural to [the string
](https://github.com/Automattic/wp-calypso/blob/trunk/client/me/purchases/manage-purchase/index.jsx#L660)
`A professionally built %(numberOfIncludedPages)d page website in 4 business days or less`

In some languages, a separate translation is needed if you have more than one page.

Also, I think there should be hyphen between the number and the word "page", such as "A professionally built 2-page website."

There is also an extra space before the colon in [this string](https://github.com/Automattic/wp-calypso/blob/trunk/client/me/purchases/manage-purchase/purchase-meta.jsx#L245):

`Service : %(oneTimeFee)s (one-time)`

Thanks!

**Comment by ddc22** [comment]
Hey @emilyaudela thanks for this feedback. 

Technically there might not be a scenario where a single page is provided as the default offering but theoretically it is a possibility. So I am going to include a version with the word "single". Just in case.
`A professionally built single page website in 4 business days or less`



The extra space before the colon is also added here which I will fix.
https://github.com/Automattic/wp-calypso/blob/ed1f87d0b8b28069df967b84326197b3878f2f63/client/me/purchases/manage-purchase/purchase-meta.jsx#L253-L254

Is there a reason or best practice for not adding a space before a colon ( English is not my first language 😄 ). I just added it because it looked more pleasing to the eye 😁

**Comment by emilyaudela** [comment]
Thanks for looking into this!

> Technically there might not be a scenario where a single page is provided as the default offering but theoretically it is a possibility. So I am going to include a version with the word "single". Just in case.
A professionally built single page website in 4 business days or less

Awesome! Thanks! 👍 

> Is there a reason or best practice for not adding a space before a colon.

In English, there aren't usually spaces before punctuation marks. 😄

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7508317).
